server/src/infra/ldap_handler.rs

```diff
@@ -211,6 +224,216 @@ fn root_dse_response(base_dn: &str) -> LdapOp {
+        let mut group_object_classes = get_default_group_object_classes();
+        group_object_classes.extend(schema.extra_group_object_classes.clone());
+
+        Self {
+            base: PublicSchema::from(schema.clone()).get_schema().clone(),
+            user_object_classes: ObjectClassList(user_object_classes),
+            group_object_classes: ObjectClassList(group_object_classes),
+        }
+    }
+
+    fn required_user_attributes(&self) -> AttributeList {
+        let attributes = self
+            .user_attributes
+            .attributes
+            .iter()
+            .filter(|&a| get_required_user_attributes().contains(&a.name))
+            .cloned()
+            .collect();
+
+        AttributeList { attributes }
+    }
+
+    fn optional_user_attributes(&self) -> AttributeList {
+        let attributes = self
+            .user_attributes
+            .attributes
+            .iter()
+            .filter(|&a| !get_required_user_attributes().contains(&a.name))
+            .cloned()
+            .collect();
+
+        AttributeList { attributes }
+    }
+
+    fn required_group_attributes(&self) -> AttributeList {
+        let attributes = self
+            .group_attributes
+            .attributes
+            .iter()
+            .filter(|&a| get_required_group_attributes().contains(&a.name))
+            .cloned()
+            .collect();
+
+        AttributeList { attributes }
+    }
+
+    fn optional_group_attributes(&self) -> AttributeList {
+        let attributes = self
+            .group_attributes
+            .attributes
+            .iter()
+            .filter(|&a| !get_required_group_attributes().contains(&a.name))
+            .cloned()
+            .collect();
+
+        AttributeList { attributes }
+    }
+
+    // See RFC4512 section 4.2.2 "attributeTypes"
+    fn formatted_attribute_list(&self) -> Vec<Vec<u8>> {
+        let mut formatted_list: Vec<Vec<u8>> = Vec::new();
+
+        for (index, attribute) in self.all_attributes().attributes.into_iter().enumerate() {
+            formatted_list.push(
+                format!(
+                    "( 2.{} NAME '{}' DESC 'LLDAP: {}' SUP {:?} )",
+                    (index + 4),
+                    attribute.name,
+                    if attribute.is_hardcoded {
+                        "builtin attribute"
+                    } else {
+                        "custom attribute"
+                    },
+                    attribute.attribute_type
+                )
+                .into_bytes()
+                .to_vec(),
+            )
+        }
+
+        formatted_list
+    }
+
+    pub fn all_attributes(&self) -> AttributeList {
+        let mut combined_attributes = self.user_attributes.attributes.clone();
+        combined_attributes.extend(self.group_attributes.attributes.clone());
+        AttributeList {
+            attributes: combined_attributes,
+        }
+    }
+}
+
+fn schema_response(schema: &Schema) -> LdapOp {
```

Comment:
Same here, we can propagate the need for a value, not a reference. And then in the callsite, it becomes obvious that we didn't need to clone at all :)
Up to you whether this should take a `PublicSchema` or not, but I think it should.

Can you expand on the name a bit? `make_ldap_subschema_entry` maybe?